File: engine/src/main/java/com/homesnap/engine/controller/light/Light.java
package com.homesnap.engine.controller.light;

/*
 * #%L
 * HomeSnapEngine
 * %%
 * Copyright (C) 2011 - 2016 A. de Giuli
 * %%
 * This file is part of HomeSnap done by Arnaud de Giuli (arnaud.degiuli(at)free.fr)
 *     helped by Olivier Driesbach (olivier.driesbach(at)gmail.com).
 * 
 *     HomeSnap is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 * 
 *     HomeSnap is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 * 
 *     You should have received a copy of the GNU General Public License
 *     along with HomeSnap. If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */

import java.util.ArrayList;
import java.util.List;

import com.homesnap.engine.controller.Controller;
import com.homesnap.engine.controller.what.impl.IntegerState;
import com.homesnap.engine.controller.what.impl.OnOffState;
import com.homesnap.engine.controller.what.impl.PercentageState;
import com.homesnap.engine.controller.what.impl.RGBState;
import com.homesnap.engine.controller.who.Who;

public class Light extends Controller {

	/** uuid */
	private static final long serialVersionUID = 1L;
	
	@Override
	public Who getWho() {
		return Who.LIGHT;
	}

	public OnOffState getStatus() {
		return (OnOffState) get(LightStateName.status.name().toLowerCase());
	}

	public void setStatus(OnOffState status) {
		set(LightStateName.status.name().toLowerCase(), status);
	}

	public PercentageState getLevel() {
		return (PercentageState) get(LightStateName.level.name());
	}
	
	public void setLevel(PercentageState value) {
		set(LightStateName.level.name(), value);
	}
	
	public OnOffState getBlink() {
		return (OnOffState) get(LightStateName.blink_time.name());
	}
	
	public void setBlink(OnOffState value) {
		set(LightStateName.blink_time.name(), value);
	}
	
	public IntegerState getBlinkTime() {
		// in millisecond
		return (IntegerState) get(LightStateName.blink.name());
	}
	
	public void setBlinkTime(IntegerState value) {
		// in millisecond
		set(LightStateName.blink.name(), value);
	}
	
	public RGBState getColor() {
		return (RGBState) get(LightStateName.color.name());
	}

	@Override
	public List<String> getStateList() {
		List<String> result = new ArrayList<String>();
		LightStateName[] list = LightStateName.values();
		for (int i = 0; i < list.length; i++) {
			result.add(list[i].name());
		};
		return result;
	}
}

File: web/src/main/webapp/scripts/snaphome/widget/light/Light.js
/*
 * #%L
 * MyDomoWebServer
 * %%
 * Copyright (C) 2011 - 2013 A. de Giuli
 * %%
 * This file is part of MyDomo done by A. de Giuli (arnaud.degiuli(at)free.fr).
 * 
 *     MyDomo is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 * 
 *     MyDomo is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 * 
 *     You should have received a copy of the GNU General Public License
 *     along with MyDomo.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */
define([
	"dojo/_base/declare",
	"dijit/form/ToggleButton",
	"dijit/form/HorizontalSlider",
	"dijit/_WidgetBase",
	"dijit/_TemplatedMixin",
	"dojo/text!./templates/Light.html",
	"dojo/dom-style",
	"dojo/_base/fx",
	"dojo/_base/lang",
	"dojo/request",
	"dojo/json",
	"snaphome/utils/Message"
], function(declare, ToggleButton, HorizontalSlider, _WidgetBase, _TemplatedMixin, template, domStyle, baseFx, lang, request, JSON, msg) {
	
	return declare([_WidgetBase, _TemplatedMixin], {
		baseClass: "light",
		templateString: template,
		button: null,
		slider: null,
		lock: false,
		adress: null,
		showSlider: false,
		switchOnOff : function(status, value) {
			var component = this;
			if (!component.lock) {
				component.lock = true;
				// Call REST API to change light
				request.put("/house/controllers/" + this.adress + "/" + status, {sync: true, handleAs: "json"})
					.then(function(data){
						if (data.status == status) { 
							if ("LIGHT_ON" == status) {
								component.set('label', 'ON');
								component.button.set('iconClass', 'lightOnIcon');
								component.slider.set('value', value);
							} else { // Off
								component.set('label', 'OFF');
								component.button.set('iconClass', 'lightOffIcon');
								component.slider.set('value', value);
							}
						} else {
							msg.displayError("Error from gateway. Status doesn't change. Try later.");
							// TODO raise an error to reset the button status to the previous value
						}
					}, function(error) {
						msg.displayError(error);
				});
				component.lock = false
			}
		},
		startup: function() {
			var component = this;
			this.button = new ToggleButton ({
				checked: false,
				iconClass: 'lightOffIcon',
				label: 'OFF',
				onChange: function(status, value){
					if (!value){
						status ? value = 100 : value = 0;
					}
					if (status) {
						component.switchOnOff("LIGHT_ON", 100);
						console.log('on');
					} else {
						component.switchOnOff("LIGHT_OFF", 0);
						console.log('off');
					}
				},
				postCreate: function() {
					this.set('showLabel', false);
				}
			}, this.id + "progButtonNode");

			if (this.showSlider) {
				this.slider = new HorizontalSlider({
					value: 0,
					minimum: 0,
					maximum: 100,
					intermediateChanges: true,
					style: "width: 300px",
					onChange: function(value){
						if (!component.lock) {
							if (value > 0) {
								component.button.onChange(true, value);
							} else {
								component.button.onChange(false, value);
							}
						}
					}
				}, this.id+"slider");
			}
		}
	});
});